x/erc20/keeper/precompile.go

```diff
@@ -0,0 +1,64 @@
+// Copyright Tharsis Labs Ltd.(Evmos)
+// SPDX-License-Identifier:ENCL-1.0(https://github.com/evmos/evmos/blob/main/LICENSE)
+
+package keeper
+
+import (
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/evmos/evmos/v15/precompiles/erc20"
+	"github.com/evmos/evmos/v15/precompiles/werc20"
+	"github.com/evmos/evmos/v15/x/erc20/types"
+)
+
+// RegisterERC20Extensions registers the ERC20 precompiles with the EVM.
+func (k Keeper) RegisterERC20Extensions(ctx sdk.Context) error {
+	precompiles := make([]vm.PrecompiledContract, 0)
+	params := k.evmKeeper.GetParams(ctx)
+	evmDenom := params.EvmDenom
+	logger := ctx.Logger()
+
+	k.IterateTokenPairs(ctx, func(tokenPair types.TokenPair) bool {
+		// skip registration if token is native or if it has already been registered
+		// NOTE: this should handle failure during the selfdestruct
+		if tokenPair.ContractOwner != types.OWNER_MODULE ||
+			params.IsPrecompileRegistered(tokenPair.Erc20Address) {
+			return false
+		}
+
+		var (
+			err        error
+			precompile vm.PrecompiledContract
+		)
+
+		if tokenPair.Denom == evmDenom {
+			precompile, err = werc20.NewPrecompile(tokenPair, k.bankKeeper, k.authzKeeper, *k.transferKeeper)
+		} else {
+			precompile, err = erc20.NewPrecompile(tokenPair, k.bankKeeper, k.authzKeeper, *k.transferKeeper)
+		}
+
+		if err != nil {
+			panic(fmt.Errorf("failed to instantiate ERC-20 precompile for denom %s: %w", tokenPair.Denom, err))
```

Comment:
should this really be panicking?

Reply:
no, let's return the error

Reply:
to return an error here we need to adjust the iteration logic here, because this function can only return bool:

```go
k.IterateTokenPairs(ctx, func(tokenPair types.TokenPair) bool {
    ...
}
```

We could just use `k.GetTokenPairs` and then iterate using a plain `for` loop. That way we could return an error here.

Reply:
@MalteHerrmann you can define the error as follows:

```
var err error
k.IterateTokenPairs(ctx, func(tokenPair types.TokenPair) bool {
    if err != nil {
        return true
    }
})

if err != nil {
   return err
}
```